docs/hardware_keyboard_guidelines.md

```diff
@@ -144,10 +144,38 @@ The `rules.mk` file can also be placed in a sub-folder, and its reading order is
         * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/rules.mk`
           * `keyboards/top_folder/keymaps/a_keymap/rules.mk`
           * `users/a_user_folder/rules.mk`
+        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_rules.mk`
+      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_rules.mk`
+    * `keyboards/top_folder/sub_1/sub_2/post_rules.mk`
+  * `keyboards/top_folder/sub_1/post_rules.mk`
+* `keyboards/top_folder/post_rules.mk`
```

Comment:
The way I read this section says that values in `keyboards/top_folder/post_rules.mk` override values in `keyboards/top_folder/sub_1/post_rules.mk`. Is that how the code actually behaves?

Reply:
I envision the following usage.

For example, let's assume that top_folder supports the following two switch macros
* FEATURES_1_ENABLE
* FEATURES_2_ENABLE

In top_folder/post_rules.mk, the following will be handled.
```makefile
ifeq ($(strip $(FEATURE_1_ENABLE)), yes)
   # Configuration for common_features.mk
endif
ifeq ($(strip $(FEATURE_2_ENABLE)), yes)
   # Configuration for common_features.mk
endif
```

Now, let's say we add top_folder/sub1, where we want to add a shortcut FEATURE_ALL_ENABLE.

The file top_folder/sub1/post_rules.mk should look like this
```makefile
ifeq ($(strip $(FEATURE_ALL_ENABLE)), yes)
   FEATURE_1_ENABLE = yes
   FEATURE_2_ENABLE = yes
endif
```

Reply:
Sorry for the delayed response here, I realized I didn't answer while doing a tab sweep.

That's not what I'm asking about here. I believe this list does not match the actual processing done here: <https://github.com/qmk/qmk_firmware/pull/8422/files/e396557341533bb77dc0b0a2b2fa21a774d898e1#diff-94923ddf56c801cf592868f3d0560d912344285dbc067dec0cfb0e2ec0034f28R325-R340>

This list says that `keyboards/top_folder/sub_1/post_rules.mk` is executed first and `keyboards/top_folder/post_rules.mk` is executed second, but the code has that order reversed.

Reply:
I'm going to load `rules.mk` and `post_rules.mk` in reverse order. Am I doing something wrong?

https://github.com/mtei/qmk_firmware/blob/e396557341533bb77dc0b0a2b2fa21a774d898e1/build_keyboard.mk#L76-L91

```makefile
# Pull in rules.mk files from all our subfolders
ifneq ("$(wildcard $(KEYBOARD_PATH_5)/rules.mk)","")
    include $(KEYBOARD_PATH_5)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_4)/rules.mk)","")
    include $(KEYBOARD_PATH_4)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_3)/rules.mk)","")
    include $(KEYBOARD_PATH_3)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_2)/rules.mk)","")
    include $(KEYBOARD_PATH_2)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_1)/rules.mk)","")
    include $(KEYBOARD_PATH_1)/rules.mk
endif
```

https://github.com/mtei/qmk_firmware/blob/e396557341533bb77dc0b0a2b2fa21a774d898e1/build_keyboard.mk#L325-L340

```makefile
# Pull in post_rules.mk files from all our subfolders
ifneq ("$(wildcard $(KEYBOARD_PATH_1)/post_rules.mk)","")
    include $(KEYBOARD_PATH_1)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_2)/post_rules.mk)","")
    include $(KEYBOARD_PATH_2)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_3)/post_rules.mk)","")
    include $(KEYBOARD_PATH_3)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_4)/post_rules.mk)","")
    include $(KEYBOARD_PATH_4)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_5)/post_rules.mk)","")
    include $(KEYBOARD_PATH_5)/post_rules.mk
endif
```

Reply:
@skullydazed do you disagree with reversing the ordering?

Reply:
Yes, I disagree. Unless I'm missing something here it seems that loading them in reverse order will mean that less specific is overriding more specific. It's not clear to me why reversing the ordering would be desirable.

Reply:
> This list says that keyboards/top_folder/sub_1/post_rules.mk is executed first and keyboards/top_folder/post_rules.mk is executed second, but the code has that order reversed.

There is no discrepancy between the description in `docs/hardware_keyboard_guidelines.md` and the behavior in `build_keyboard.mk`.

```
        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_rules.mk`
      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_rules.mk`
    * `keyboards/top_folder/sub_1/sub_2/post_rules.mk`
  * `keyboards/top_folder/sub_1/post_rules.mk`
* `keyboards/top_folder/post_rules.mk`
* `common_features.mk`
```

```
# Pull in post_rules.mk files from all our subfolders
 ifneq ("$(wildcard $(KEYBOARD_PATH_1)/post_rules.mk)","")
     include $(KEYBOARD_PATH_1)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_2)/post_rules.mk)","")
     include $(KEYBOARD_PATH_2)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_3)/post_rules.mk)","")
     include $(KEYBOARD_PATH_3)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_4)/post_rules.mk)","")
     include $(KEYBOARD_PATH_4)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_5)/post_rules.mk)","")
     include $(KEYBOARD_PATH_5)/post_rules.mk
 endif
```

Reply:
> Yes, I disagree. Unless I'm missing something here it seems that loading them in reverse order will mean that less specific is overriding more specific. It's not clear to me why reversing the ordering would be desirable.

Are you objecting to the following order of execution?

```
        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_rules.mk`
      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_rules.mk`
    * `keyboards/top_folder/sub_1/sub_2/post_rules.mk`
  * `keyboards/top_folder/sub_1/post_rules.mk`
* `keyboards/top_folder/post_rules.mk`
* `common_features.mk`
```

Reply:
As far as I can tell, yes, that was the dispute.

As a somewhat hypothetical example, three systems `ABC`, `DEF`, and `GHI` all requiring `CANBUS`.

Reversed:
* sub_4: 
  * `if ABC=yes then NEEDS_CANBUS=yes; SRC += abc.c` 
  * `if DEF=yes then NEEDS_CANBUS=yes; SRC += def.c`
  * `if GHI=yes then NEEDS_CANBUS=yes; SRC += ghi.c`
* sub_3:
  * `if NEEDS_CANBUS=yes AND PLATFORM=CHIBIOS then SRC += canbus_chibios.c`
  * `if NEEDS_CANBUS=yes AND PLATFORM=AVR then SRC += canbus_avr.c`

Sequential:
* sub_3:
  * `if (ABC=yes OR DEF=yes OR GHI=yes) AND PLATFORM=CHIBIOS  then SRC += canbus_chibios.c`
  * `if (ABC=yes OR DEF=yes OR GHI=yes) AND PLATFORM=AVR then SRC += canbus_avr.c`
* sub_4:
  * `if ABC=yes then SRC += abc.c`, 
  * `if DEF=yes then SRC += def.c`
  * `if GHI=yes then SRC += ghi.c`

So the question becomes: should all intermediate levels need to check specific overall features, or can required systems be set to yes/no instead and higher levels check for dependencies instead of features?

Reply:
In that specific example shouldn’t NEED_CANBUS be set in rules.mk and then the SRC can be set in post, thus being able to use the same code on both?

(Apologies for the brevity, I’m on mobile)

Reply:
I talked through this a bit with @tzarc on discord, and the conclusion was the current ordering will be fine. I have concerns about the support load if this feature is used a lot, but if it is not the order is arbitrary.